### ISSUE TITLE
Basic unit test support for C++ libraries and executables

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle-native#31]
Add some simple and hard-coded support for implementing and running unit tests for C++ libraries and executables.

Make similar assumptions as for #2 and also:

- A single unit test suite located at a hard-coded location, implemented in C++.
- No specific test framework support, just link the test suite into an executable and run it.

**Comment by lacasseio** [comment]
The most basic test suite is a simple executable that exit with an error code when it doesn't pass. No need for a testing framework at this stage.

**Comment by adammurdoch** [comment]
Absolutely. We do need to have a plan for how a test framework would be introduced though.

**Comment by lacasseio** [comment]
We should definitely have one. One thing to keep in mind is how to have multiple test binaries for a single project. We could map the test framework and test type ([small, medium and large - an example from Google](https://testing.googleblog.com/2010/12/test-sizes.html)) as dimensions of the test component. This way we could have multiple test executables per-project.

**Comment by adammurdoch** [comment]
From a modelling perspective, I think we want to end up with something like what we already have in the software model:

- A project may have zero or more test suites.
- Each test suite is-a native component, and so has variants.
- Each test suite has-a test framework associated with it.
- "link and run an executable" is-a kind of test framework
- The variants of a test suite may or may not be the same as the variants of the code under test, eg I may run the unit tests on x64 linux only but run the functional tests on every platform, or I don't have a 'release' build type for my tests, or I run the functional tests on windows 8 and windows 10 but I only build on windows 8.

**Comment by adammurdoch** [comment]
> We could map the test framework and test type as dimensions of the test component

These are different test suites in my view.

**Comment by lacasseio** [comment]
The question would be, if we were doing the same in Java, would we implement it the same way? We don't want to diverge again from the current ecosystem, but I understand that the current Java implementation for multiple tests "executable" may need some work due to current limitation.
